ports/esp32/network_ppp.c

```diff
@@ -0,0 +1,228 @@
+#include "lwip/pppapi.h"
+
+
+typedef struct _ppp_if_obj_t {
+    mp_obj_base_t base;
+    bool initialized;
+    bool active;
+    bool connected;
+    ppp_pcb *pcb;
+    mp_obj_t stream;
+    struct netif pppif;
+    SemaphoreHandle_t inactiveWaitSem;
+    TaskHandle_t client_task_handle;
+} ppp_if_obj_t;
+
+const mp_obj_type_t ppp_if_type;
+STATIC ppp_if_obj_t ppp_obj = {{&ppp_if_type}, false, false, false, NULL, NULL};
+
+/* PPP status callback example */
+static void ppp_status_cb(ppp_pcb *pcb, int err_code, void *ctx) // wait for
+{
+    ppp_if_obj_t* self = &ppp_obj;
+    struct netif *pppif = ppp_netif(self->pcb);
+    (void)pppif;
+    LWIP_UNUSED_ARG(ctx);
+
+    switch (err_code) {
+        case PPPERR_NONE: {
+                self->connected = (pppif->ip_addr.u_addr.ip4.addr != 0);
+                break;
+            }
+        case PPPERR_USER: {
+                xSemaphoreGive(self->inactiveWaitSem);
+                break;
+            }
+        case PPPERR_CONNECT: {
+                self->connected = false;
+                break;
+            }
+        default: {
+                break;
+            }
+    }
+}
+
+STATIC mp_obj_t get_ppp(mp_obj_t stream) {
+    ppp_if_obj_t* self = &ppp_obj;
+
+    if (self->initialized) {
+        return MP_OBJ_FROM_PTR(&ppp_obj);
+    }
+    mp_get_stream_raise(stream, MP_STREAM_OP_READ | MP_STREAM_OP_WRITE);
+    self->stream = stream;
+
+    self->inactiveWaitSem = xSemaphoreCreateBinary();
+    assert(self->inactiveWaitSem != NULL);
+    self->initialized = true;
+    return MP_OBJ_FROM_PTR(&ppp_obj);
+}
+MP_DEFINE_CONST_FUN_OBJ_1(get_ppp_obj, get_ppp);
+
+static u32_t ppp_output_callback(ppp_pcb *pcb, u8_t *data, u32_t len, void *ctx)
+{
+    ppp_if_obj_t* self = &ppp_obj;
+    int err;
+    return mp_stream_rw(self->stream, data, len, &err, MP_STREAM_RW_WRITE);
+}
+
+static void pppos_client_task() {
+    ppp_if_obj_t* self = &ppp_obj;
+    uint8_t buf[256];
+
+    int err;
+    const mp_stream_p_t *stream = mp_get_stream_raise(self->stream, MP_STREAM_OP_READ);
+
+    while (ulTaskNotifyTake(pdTRUE, 0) == 0) {
+        int len = mp_stream_rw(self->stream, buf, sizeof(buf), &err, 0);
+        if (len > 0) {
+            pppos_input_tcpip(self->pcb, (u8_t *)buf, len);
+        }
+    }
+    vTaskDelete(NULL);
+}
+
+STATIC mp_obj_t ppp_active(size_t n_args, const mp_obj_t *args) {
+    ppp_if_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+
+    if (n_args > 1) {
+        if (mp_obj_is_true(args[1])) {
+            if (self->active) {
+                return mp_const_true;
+            }
+
+            self->pcb = pppapi_pppos_create(&self->pppif, ppp_output_callback, ppp_status_cb, NULL);
+
+            if (self->pcb == NULL) {
+                mp_raise_msg(&mp_type_RuntimeError, "init failed");
+            }
+            pppapi_set_default(self->pcb);
+            pppapi_connect(self->pcb, 0);
+
+            xTaskCreate(pppos_client_task, "ppp", 1024, NULL, 1, &self->client_task_handle);
```

Comment:
NULL can be replaced with `self` to pass through the PPP instance

Reply:
Done